internal/sbi/httpcallback/api_handle_dereg_notification.go

```diff
@@ -0,0 +1,138 @@
+	var doUecmDereg bool = true
+
+	requestBody, err := c.GetRawData()
+	if err != nil {
+		logger.CallbackLog.Errorf("Get Request Body error: %+v", err)
+		problemDetails := models.ProblemDetails{
+			Title:  "System failure",
+			Status: http.StatusInternalServerError,
+			Detail: err.Error(),
+			Cause:  "SYSTEM_FAILURE",
+		}
+		c.JSON(http.StatusInternalServerError, problemDetails)
+		return
+	}
+
+	err = openapi.Deserialize(&deregData, requestBody, "application/json")
+	if err != nil {
+		problemDetails := "[Request Body] " + err.Error()
+		rsp := models.ProblemDetails{
+			Title:  "Malformed request syntax",
+			Status: http.StatusBadRequest,
+			Detail: problemDetails,
+		}
+		logger.CallbackLog.Errorln(problemDetails)
+		c.JSON(http.StatusBadRequest, rsp)
+		return
+	}
+
+	ueid := c.Param("ueid")
+	amfSelf := amf_context.GetSelf()
+	ue, ok := amfSelf.AmfUeFindByUeContextID(ueid)
+	if !ok {
+		logger.CallbackLog.Errorf("AmfUe Context[%s] not found", ueid)
+		problemDetails := &models.ProblemDetails{
+			Status: http.StatusNotFound,
+			Cause:  "CONTEXT_NOT_FOUND",
+		}
+		c.JSON(http.StatusNotFound, problemDetails)
+		return
+	}
+
+	switch deregData.DeregReason {
+	case models.DeregistrationReason_UE_INITIAL_REGISTRATION:
+		// TS 23.502 - 4.2.2.2.2 General Registration
+		// Invokes the Nsmf_PDUSession_ReleaseSMContext for the corresponding access type
+		ue.SmContextList.Range(func(key, value interface{}) bool {
+			smContext := value.(*amf_context.SmContext)
+
+			if smContext.AccessType() == deregData.AccessType {
+				var problemDetails *models.ProblemDetails
+				problemDetails, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
+				if problemDetails != nil {
+					ue.GmmLog.Errorf("Release SmContext Failed Problem[%+v]", problemDetails)
+				} else if err != nil {
+					ue.GmmLog.Errorf("Release SmContext Error[%v]", err.Error())
+				}
+			}
+
+			return true
+		})
+	case models.DeregistrationReason_SUBSCRIPTION_WITHDRAWN:
+		// The notification is triggered by the new AMF executing UECM registration,
+		// UDM will use the new data to replace the old context.
+		// Therefore old AMF doesn't need to execute UECM de-registration to clean the old context stored in UDM.
+		doUecmDereg = false
+	}
+	// TS 23.502 - 4.2.2.2.2 General Registration - 14e
+	// TODO: (R16) If old AMF does not have UE context for another access type (i.e. non-3GPP access),
+	// the Old AMF unsubscribes with the UDM for subscription data using Nudm_SDM_unsubscribe
+	if ue.SdmSubscriptionId != "" {
+		var problemDetails *models.ProblemDetails
+		problemDetails, err = consumer.SDMUnsubscribe(ue)
+		if problemDetails != nil {
+			logger.GmmLog.Errorf("SDM Unubscribe Failed Problem[%+v]", problemDetails)
+		} else if err != nil {
+			logger.GmmLog.Errorf("SDM Unubscribe Error[%+v]", err)
+		}
+		ue.SdmSubscriptionId = ""
+	}
+
+	if doUecmDereg {
+		// Use old AMF as the backup AMF
+		backupAmfInfo := models.BackupAmfInfo{
```

Comment:
Now there is no config for backup amf, so don't fill backup amf info to UDM for now.

Reply:
Ok, I will remove it.